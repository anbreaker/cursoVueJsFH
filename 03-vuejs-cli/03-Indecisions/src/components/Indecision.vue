<template>
  <div>
    <h1>Indecision</h1>
  </div>
</template>

<script>
  export default {
    
  }
</script>

<style scoped>

</style>