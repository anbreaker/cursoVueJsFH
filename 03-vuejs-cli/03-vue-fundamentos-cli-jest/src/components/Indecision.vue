<template>
  <div>
    <img v-if="img" :src="img" alt="bg" />
    <div class="bg-dark"></div>

    <div class="indecision-container">
      <input type="text" placeholder="Ask me a Question!" v-model="question" />
      <h3>Remember to end with a question mark (?).</h3>
      <div v-if="isValidQuestion">
        <h2>{{ question }}</h2>
        <h1>{{ answer }}</h1>
      </div>
    </div>
  </div>
</template>

<script>
  export default {
    data() {
      return {
        question: null,
        answer: null,
        img: null,
        isValidQuestion: false,
      };
    },

    methods: {
      async getAnswer() {
        try {
          const { answer, image } = await fetch('https:/yesno.wtf/api').then((res) =>
            res.json()
          );

          this.answer = answer;
          this.img = image;

          console.log(answer, image);
        } catch (error) {
          this.img = null;
          this.answer = '404 Not found API';

          console.log('IndecisionComponent: ', error);
        }
      },
    },

    watch: {
      question(newValue) {
        this.isValidQuestion = false;

        console.log({ newValue });

        if (!newValue.includes('?')) return;

        console.log({ newValue });

        this.isValidQuestion = true;

        this.getAnswer();
      },
    },
  };
</script>

<style scoped>
  img,
  .bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
  }

  .bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
  }

  .indecision-container {
    position: relative;
    z-index: 99;
  }

  input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
  }

  input:focus {
    outline: none;
  }

  h1,
  h2,
  h3 {
    color: white;
  }

  h2 {
    margin-top: 150px;
  }
</style>
