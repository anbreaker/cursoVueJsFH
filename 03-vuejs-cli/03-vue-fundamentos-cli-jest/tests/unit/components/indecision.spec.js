import { shallowMount } from '@vue/test-utils';
import Indecision from '@/components/Indecision';
describe('Indecision Component', () => {
  let wrapper;
  let consoleSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: 'yes',
          forced: false,
          image: 'https://yesno.wtf/assets/yes/2.gif',
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);

    consoleSpy = jest.spyOn(console, 'log');

    jest.clearAllMocks();
  });

  test('Should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Write in the input should not execute anything (console.log)', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

    const input = wrapper.find('input');

    await input.setValue('Hello');

    expect(consoleSpy).toHaveBeenCalledTimes(1);

    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test('typing "?" should fire the getAnswer', async () => {
    try {
      const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

      const input = wrapper.find('input');

      await input.setValue('Hello?');

      expect(consoleSpy).toHaveBeenCalledTimes(2);

      expect(getAnswerSpy).toHaveBeenCalled();
    } catch (error) {
      console.log(error);
    }
  });

  test('testing in getAnswer', async () => {
    await wrapper.vm.getAnswer();

    const img = wrapper.find('img');
    expect(img.exists()).toBe(true);
    expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif');
    expect(wrapper.vm.answer).toBe('yes');
  });

  test('tests on getAnswer with API failure', async () => {
    // TODO Failure API
    fetch.mockImplementationOnce(() => Promise.reject('API is Down'));

    await wrapper.vm.getAnswer();

    const img = wrapper.find('img');

    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe('404 Not found API');
  });
});
