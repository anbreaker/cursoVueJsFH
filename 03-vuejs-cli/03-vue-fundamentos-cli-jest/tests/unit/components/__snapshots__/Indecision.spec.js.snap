// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Should match with the snapshot 1`] = `
<div>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Ask me a Question!">
    <h3>Remember to end with a question mark (?).</h3>
    <!--v-if-->
  </div>
</div>
`;
